lib/devise_invitable/controllers/helpers.rb

```diff
@@ -3,19 +3,20 @@ module DeviseInvitable::Controllers::Helpers
 
   included do
   end
-  
+
   def after_invite_path_for(inviter, invitee = nil)
-    after_sign_in_path_for(inviter)
+    signed_in_root_path(inviter)
   end
-  
+
   def after_accept_path_for(resource)
-    after_sign_in_path_for(resource)
```

Comment:
I don't think this should have changed. Devise redirects you to `after_sign_in_path_for` when a new user is registered; I think the equivalent makes sense here.  https://github.com/heartcombo/devise/blob/acc45c5a44c45b252ccba65fd169a45af73ff369/app/controllers/devise/registrations_controller.rb#L114

Reply:
> I think the equivalent makes sense here. 

Can you elaborate? Did you read #733?

### What is the effect of this change?

The change you pointed out replaces `after_sign_in_path_for` with `signed_in_root_path`. As you can see, [`#after_sign_in_path_for` just calls out to `#signed_in_root_path` anyway](https://github.com/heartcombo/devise/blob/master/lib/devise/controllers/helpers.rb#L215-L217), with a short-circuit operator in front:

```ruby
def after_sign_in_path_for(resource_or_scope)
  stored_location_for(resource_or_scope) || signed_in_root_path(resource_or_scope)
end
```

Thus, the intended effect is to **skip this short-circuit branch**.

#### Caveat

As it turns out, there is also an unintended side effect: before, if you explicitly overrode `#after_sign_in_path_for`, that would implicitly also override `#after_accept_path_for`. Now, you would need to manually override both.

Is this the problem?

### What is `#stored_location_for`, and why should we skip it?

Imagine you weren't logged in, and you opened a link to an auth-required page. Devise does the following:

1. remember your destination,
2. redirect you to the login page, and
3. send you back to the path from Step 1 after login.

Steps 1 and 3 depend on a stored location (hence, `#stored_location_for`). AFAIK, this is the only user flow that requires this method. It does not apply to invitations: people only arrive at invitations by following a link in their email, so there's no "original destination" to redirect them back to.

Seems harmless enough, but now imagine the following scenario: I don't have an account on your site, and I try to visit an auth-required page. No dice! I can't create an account either, because your site is invitation-only. So, I wait around for a couple days, get an invitation from a friend, and then set up my account. Once I log in, I'm automatically redirected back to the page I tried to visit a few days back, instead of to the new-user landing page. Not ideal.

---

Do you have another approach to solving this problem?

Reply:
I did read #733, but I don't agree with it, at least in the context of what happens after you accept an invite (I'm not fussed on what happens after you *send* an invite).

The reason you should call `after_sign_in_path_for` is because Devise recommends you override it to add custom behaviour: https://github.com/heartcombo/devise/wiki/How-To:-Redirect-to-a-specific-page-on-successful-sign-in. So I agree with your caveat; the problem is your caveat is much more common than you imply.

> Seems harmless enough, but now imagine the following scenario: I don't have an account on your site, and I try to visit an auth-required page. No dice! I can't create an account either, because your site is invitation-only. So, I wait around for a couple days, get an invitation from a friend, and then set up my account. Once I log in, I'm automatically redirected back to the page I tried to visit a few days back, instead of to the new-user landing page. Not ideal.

This feels like a pretty niche edge case to me. Regardless, it's better resolved by overriding `after_sign_in_path_for` anyway.

Reply:
### Proposed solution

```ruby
# this, scoped to invitation controllers only?
def after_sign_in_path_for(resource_or_scope)
  signed_in_root_path(resource_or_scope)
end

# instead of this
def after_invite_path_for(inviter)
  signed_in_root_path(inviter)
end

def after_accept_path_for(resource)
  signed_in_root_path(resource)
end
```

### Rationale / Discussion

> your caveat is much more common than you imply.

Sorry, I didn't mean to understate the importance of this side effect. I 100% agree that it's a problem that deserves fixing—Devise Invitable extends Devise, and this wrinkle reduces their compatibility, which is no good.

> This feels like a pretty niche edge case to me.

It's precisely _because_ it's an edge case that it should work sanely out of the box. It's an undue burden on the users of a library to have to seek out and patch up all the little edge cases that work contrary to expectation. Redirecting to a page you forgot you tried to visit last week is bonkers default behavior, not to mention being almost impossible to discover on your own.

Reply:
For what it's worth, `stored_location_for` lives [in the `session`](https://github.com/heartcombo/devise/blob/715192a7709a4c02127afb067e66230061b82cf2/lib/devise/controllers/store_location.rb#L19) which is only valid until the [browser closes](https://stackoverflow.com/a/1452927/641293)

Reply:
I believe the browser setting for persisting tabs across restarts prevents sessions from being reset.

No one I know has this setting disabled, meaning that in practice, sessions never die.